File: examples/e2e-travel-concierge-py/agent/chat/providers/openai_shared.py
# Shared chat-completion logic for the OpenAI-compatible providers (plain
# OpenAI and Azure OpenAI), Python port of ../../openaiShared.ts. AzureOpenAI
# extends OpenAI in the `openai` package too, so the same code drives both --
# only the client instance and the model/deployment name differ.

from __future__ import annotations

import json
from typing import Any, List

from agent.chat.llm_error import llm_error_from_sdk_error
from agent.chat.tool_schemas import SYSTEM_PROMPT, TOOL_SCHEMAS
from agent.chat.types import LLMMessage, TextResponse, ToolCallRequest


def chat_complete(client: Any, model: str, messages: List[LLMMessage], provider_label: str) -> Any:
    try:
        res = client.chat.completions.create(
            model=model,
            messages=_to_openai_messages(messages),
            tools=[
                {"type": "function", "function": {"name": t.name, "description": t.description, "parameters": t.parameters}}
                for t in TOOL_SCHEMAS
            ],
        )
    except Exception as err:  # noqa: BLE001
        raise llm_error_from_sdk_error(provider_label, err) from err

    choice = res.choices[0].message if res.choices else None
    tool_calls = getattr(choice, "tool_calls", None) if choice else None
    tool_call = tool_calls[0] if tool_calls else None
    if tool_call is not None and tool_call.type == "function":
        return ToolCallRequest(
            tool_call_id=tool_call.id,
            tool_name=tool_call.function.name,
            args=json.loads(tool_call.function.arguments or "{}"),
        )

    return TextResponse(content=(choice.content if choice and choice.content else "") or "")


def _to_openai_messages(messages: List[LLMMessage]) -> List[dict]:
    mapped: List[dict] = [{"role": "system", "content": SYSTEM_PROMPT}]
    for m in messages:
        if m.role == "tool":
            mapped.append({"role": "tool", "tool_call_id": m.tool_call_id or "", "content": m.content})
        elif m.role == "assistant" and m.tool_call_id:
            mapped.append(
                {
                    "role": "assistant",
                    "content": None,
                    "tool_calls": [
                        {
                            "id": m.tool_call_id,
                            "type": "function",
                            "function": {"name": m.tool_name or "", "arguments": json.dumps(m.tool_args or {})},
                        }
                    ],
                }
            )
        elif m.role == "assistant":
            mapped.append({"role": "assistant", "content": m.content})
        else:
            mapped.append({"role": "user", "content": m.content})
    return mapped

File: examples/e2e-travel-concierge/agent/chat/providers/openaiShared.ts
// Shared chat-completion logic for the OpenAI-compatible providers (plain OpenAI
// and Azure OpenAI). AzureOpenAI extends OpenAI, so the same code drives both —
// only the client instance and the model/deployment name differ.
import type OpenAI from 'openai';
import { SYSTEM_PROMPT, TOOL_SCHEMAS } from '../toolSchemas.js';
import { llmErrorFromSdkError } from './llmError.js';
import type { LLMMessage, LLMResponse } from './types.js';

export async function chatComplete(
  client: OpenAI,
  model: string,
  messages: LLMMessage[],
  providerLabel: string,
): Promise<LLMResponse> {
  const res = await client.chat.completions
    .create({
      model,
      messages: toOpenAIMessages(messages),
      tools: TOOL_SCHEMAS.map((t) => ({
        type: 'function' as const,
        function: { name: t.name, description: t.description, parameters: t.parameters },
      })),
    })
    .catch((err: unknown) => {
      throw llmErrorFromSdkError(providerLabel, err);
    });

  const choice = res.choices[0]?.message;
  const toolCall = choice?.tool_calls?.[0];
  if (toolCall && toolCall.type === 'function') {
    return {
      type: 'tool_call',
      toolCallId: toolCall.id,
      toolName: toolCall.function.name,
      args: JSON.parse(toolCall.function.arguments || '{}') as Record<string, unknown>,
    };
  }

  return { type: 'text', content: choice?.content ?? '' };
}

export function toOpenAIMessages(messages: LLMMessage[]): OpenAI.Chat.ChatCompletionMessageParam[] {
  const mapped: OpenAI.Chat.ChatCompletionMessageParam[] = [
    { role: 'system', content: SYSTEM_PROMPT },
  ];
  for (const m of messages) {
    if (m.role === 'tool') {
      mapped.push({ role: 'tool', tool_call_id: m.toolCallId ?? '', content: m.content });
    } else if (m.role === 'assistant' && m.toolCallId) {
      mapped.push({
        role: 'assistant',
        content: null,
        tool_calls: [
          {
            id: m.toolCallId,
            type: 'function',
            function: { name: m.toolName ?? '', arguments: JSON.stringify(m.toolArgs ?? {}) },
          },
        ],
      });
    } else if (m.role === 'assistant') {
      mapped.push({ role: 'assistant', content: m.content });
    } else {
      mapped.push({ role: 'user', content: m.content });
    }
  }
  return mapped;
}
